trackintel/preprocessing/staypoints.py

```diff
@@ -78,34 +78,42 @@ def cluster_staypoints(staypoints, method='dbscan',
                 place_id_counter = place_id_counter + max_label + 1
             
             # add staypoint - place matching to original staypoints
-            staypoints.loc[user_staypoints.index, 'place_id'] = labels
+            ret_staypoints.loc[user_staypoints.index, 'place_id'] = labels
+
 
-            
-    
         # create places as grouped staypoints
-        grouped_df = staypoints.groupby(['user_id', 'place_id'])
-        for combined_id, group in grouped_df:
-            user_id, place_id = combined_id
-            group.set_geometry(staypoints.geometry.name, inplace=True)
-    
-            if int(place_id) != -1:
-                ret_place = {}
-                ret_place['user_id'] = user_id
-                ret_place['place_id'] = place_id
-                
-                # point geometry of place
-                ret_place['center'] = Point(group.geometry.x.mean(),
-                     group.geometry.y.mean())
-                # polygon geometry of place
-                ret_place['extent'] = MultiPoint(points=list(group.geometry)).convex_hull
-    
-                ret_places = ret_places.append(ret_place, ignore_index=True)
-    
-        ret_places = gpd.GeoDataFrame(ret_places, geometry='center',
-                                      crs=staypoints.crs)
-        ret_places['place_id'] = ret_places['place_id'].astype('int')
+        temp_sp = ret_staypoints[['user_id', 'place_id', ret_staypoints.geometry.name]]
+        ret_loc = temp_sp.dissolve(by=['user_id', 'place_id'],as_index=False)
+        # filter outlier
+        ret_loc = ret_loc.loc[ret_loc['place_id'] != -1]
+        
+        # locations with only one staypoints is of type "Point"
+        point_idx = ret_loc.geom_type == 'Point'
+        ret_loc['center'] = 0 # initialize
+        ret_loc.loc[point_idx, 'center'] = ret_loc.loc[point_idx, 'geom']
+        # locations with multiple staypoints is of type "MultiPoint"
+        ret_loc.loc[~point_idx, 'center'] = ret_loc.loc[~point_idx, 'geom'].apply(lambda p: Point(np.array(p)[:,0].mean(), 
+                                                                                                np.array(p)[:,1].mean()))
+        
+        # extent is the convex hull of the geometry
+        ret_loc['extent'] = ret_loc['geom'].apply(lambda p: p.convex_hull)                                                                                        
+        # convex_hull of one point would be a Point and two points a Linestring, 
+        # we change them into Polygon by creating a buffer of epsilon around them.
+        pointLine_idx = (ret_loc['extent'].geom_type == 'LineString') | (ret_loc['extent'].geom_type == 'Point')
+        
+        # TODO: how to know if the crs is geodetic or projection?
```

Comment:
I would say that we don't do any projections here. We should expect the user to know the data and then the epsilon that we use for the buffer should be the same unit as we use it in DBScan.
I think at the moment there is still a mistake in the code if I am not getting this wrong. 
Example: 
At the moment it could be the case that if the data is in WGS84 and the user _wants_ to cluster the data in degree (for what ever reason...), he would pass _epsilon=0.01_ to the method. The places then get clustered using epsilon=0.01 degree.
For the extent of places with only one point you read the 0.01 degrees as meters and use the meters_to_decimal_degrees function to transform it into something very small...
You could use this approach if the user adds the haversine keyword for the distance matrix metric as we can then expect that the distance is in meters.